zrml/neo-swaps/src/types/liquidity_tree.rs

```diff
@@ -0,0 +1,1205 @@
+/// what its fees are, propagating fees to this node takes `O(depth)` operations (or, equivalently,
+/// `O(log_2(node_count))`).
+///
+/// # Attributes
+///
+/// - `nodes`: A vector which holds the nodes of the tree. The nodes are ordered by depth (the root
+///   is the first element of `nodes`) and from left to right. For example, the right-most
+///   grandchild of the root is at index `6`.
+/// - `account_to_index`: Maps an account to the node that belongs to it.
+/// - `abandoned_nodes`: A vector that contains the indices of abandoned nodes.
+///
+/// # Generics
+///
+/// - `T`: The pallet configuration.
+/// - `U`: A getter for the maximum depth of the tree. Using a depth larger than `31` will result in
+///   undefined behavior.
+#[derive(Decode, Encode, Eq, MaxEncodedLen, RuntimeDebugNoBound, TypeInfo)]
+#[scale_info(skip_type_params(T, U))]
+pub struct LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    pub(crate) nodes: BoundedVec<Node<T>, LiquidityTreeMaxNodes<U>>,
+    pub(crate) account_to_index: BoundedBTreeMap<T::AccountId, u32, LiquidityTreeMaxNodes<U>>,
+    pub(crate) abandoned_nodes: BoundedVec<u32, LiquidityTreeMaxNodes<U>>,
+}
+
+// Boilerplate implementation because Rust is confused by the generic parameter `U`.
+impl<T, U> Clone for LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    fn clone(&self) -> Self {
+        LiquidityTree {
+            nodes: self.nodes.clone(),
+            account_to_index: self.account_to_index.clone(),
+            abandoned_nodes: self.abandoned_nodes.clone(),
+        }
+    }
+}
+
+// Boilerplate implementation because Rust is confused by the generic parameter `U`.
+impl<T, U> PartialEq for LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.nodes == other.nodes
+            && self.account_to_index == other.account_to_index
+            && self.abandoned_nodes == other.abandoned_nodes
+    }
+}
+
+impl<T, U> LiquidityTree<T, U>
+where
+    T: Config,
+    U: Get<u32>,
+{
+    /// Create a new liquidity tree.
+    ///
+    /// Parameter:
+    ///
+    /// - `account`: The account to which the tree's root belongs.
+    /// - `stake`: The stake of the tree's root.
+    pub(crate) fn new(
+        account: T::AccountId,
+        stake: BalanceOf<T>,
+    ) -> Result<LiquidityTree<T, U>, DispatchError> {
+        let root = Node::new(account.clone(), stake);
+        let nodes = vec![root]
+            .try_into()
+            .map_err(|_| LiquidityTreeError::AccountNotFound.into_dispatch::<T>())?;
+        let mut account_to_index: BoundedBTreeMap<_, _, _> = Default::default();
+        account_to_index
+            .try_insert(account, 0u32)
+            .map_err(|_| LiquidityTreeError::AccountNotFound.into_dispatch::<T>())?;
+        let abandoned_nodes = Default::default();
+        Ok(LiquidityTree { nodes, account_to_index, abandoned_nodes })
+    }
```

Comment:
3ed2c29d95f1821cdac2783b8e32832fc1eb38ae

Reply:
Keep in mind that if you do name the errors explicitly, it's easier to debug the cause of the error. So I would suggest to define distinct errors.

Reply:
or you can use log::error before each one

Reply:
> These errors only occur in case of a misconfiguration, which will be caught during testing, or due to a logic error.

Still, if you test and get this `StorageOverflow` error, you would need to debug to find out where this error was thrown. If you log, you will hopefully see where it happened. If you define distinct errors, you are clear about where it was thrown.

Reply:
> If you define distinct errors, you are clear about where it was thrown.

Only if you use each error exactly once.

Reply:
> Only if you use each error exactly once.

I don't get this. My suggestion is to define: `NodesStorageOverflow` and `AccountToIndexStorageOverflow`. Those are distinct errors. If you just define `StorageOverflow` twice in the same function, you won't know easily which error was thrown if it happens.

Reply:
20f2b50f599ba7aacf08ffe75d459099b740c35f

> Still, if you test and get this StorageOverflow error, you would need to debug to find out where this error was thrown.

No, the backtrace gives you that information.

> If you just define StorageOverflow twice in the same function, you won't know easily which error was thrown if it happens.

If you use the same error twice anywhere in the code, you won't easily know from where it was thrown without a backtrace.